### PR TITLE
[RLlib] New API stack: Unify get/set_state, save, restore, from_checkpoint for RLModule classes.

### PR DESCRIPTION
**Pull request by sven1977**
New API stack: Unify get/set_state, save, restore, from_checkpoint for all components.
* All new API stack components: Algorithm, EnvRunner, LearnerGroup, Learner, and RLModule should abide to the exact same checkpointing APIs.
* Meaning, they should all support:
* * get_state(components=None) and set_state, where components allows filtering for only certain aspects of the state (e.g. `EnvRunner.get_state(components="env_to_module_connector")` to save the time compiling possibly expensive model weights, which may not be needed by the caller).
* * `save(path=None) -> Checkpoint` and `restore(checkpoint) -> None`, similar to how Algorithm/tune.Trainable already works.
* * `[cls].from_checkpoint(cls, checkpoint, config=None)`. Here, an optional config can be passed to override the (pickled) checkpointed one. This pattern will also help when transferring pickle-based checkpoints from one python version into a higher one via a msgpack intermediary checkpoint (which won't have the config information as this is usually NOT serializable).
* * TODO: `save_as_msgpack(path=None) -> Checkpoint`.

This PR only covers the RLModule classes (incl. MultiAgentRLModule) and updates the docs accordingly.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/rllib/rllib-rlmodule.rst [hunk 1]
```diff
@@ -31,33 +36,60 @@ RL Modules (Alpha)
...
+RL Module is a neural network container that implements three public methods:
+:py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_train`,
+:py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_exploration`,
+and :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_inference`.
+Each method corresponds to a distinct reinforcement learning and/or production phase.
 
+The :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_inference` method computes
+actions in a purely greedy fashion. It is normally used only in production settings (after
+training has been completed) or for evaluation purposes parallel to training.
+RLlib's :py:class:`~ray.rllib.evaluation.env_runner.EnvRunner` classes will chose this method
```
Comment (simonsays1980) [inline doc/source/rllib/rllib-rlmodule.rst:70]: chose -> choose

File: doc/source/rllib/rllib-rlmodule.rst [hunk 2]
```diff
@@ -31,33 +36,60 @@ RL Modules (Alpha)
...
+is set to False.
 
-RL Module is a neural network container that implements three public methods: :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_train`, :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_exploration`, and :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_inference`. Each method corresponds to a distinct reinforcement learning phase.
+:py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_exploration` computes actions
+in an exploratory way, balancing exploration and exploitation. It is the method normally used
+for training data collection by RLlib's :py:class:`~ray.rllib.evaluation.env_runner.EnvRunner` classes
+(if the `explore` flag in the :py:class:`~ray.rllib.algorithms.algorithm_config.AlgorithmConfig`
+is set to True, which it is by default).
+However, note that this method can also be used after training in an inference-only phase, if the user
+explicitly want the computed actions to be exploratory. Some environments - like the rock-paper-scissors game -
```
Comment (simonsays1980) [inline doc/source/rllib/rllib-rlmodule.rst:80]: Would put the last sentence in brackets "... (some environments ....)". It is not of direct interest at this point - it's an additional info for the reader asking for the "why".

Comment (simonsays1980) [reply doc/source/rllib/rllib-rlmodule.rst:80]: Probably adding something like "i.e. this implements a stochastic policy during inference".

File: doc/source/rllib/rllib-rlmodule.rst [hunk 3]
```diff
@@ -31,33 +36,60 @@ RL Modules (Alpha)
...
+can only be optimally solved by acting stochastically.
 
-:py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_exploration` handles acting and data collection, balancing exploration and exploitation. On the other hand, the :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_inference` serves the learned model during evaluation, often being less stochastic.
+Finally, the :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_train` method is called
+during the training update phase of an Algorithm and not only should it compute actions (or action logits/probabilities),
+but anything that is relevant and needed for the algorithm's loss function, such as computing
+Q-values in a DQN model or value function predictions in a PPO model.
```
Comment (simonsays1980) [inline doc/source/rllib/rllib-rlmodule.rst:86]: would rather name it "target Q-values in a DQN ..." as for exploration and inference we nee Q-values, too.

File: doc/source/rllib/rllib-rlmodule.rst [hunk 4]
```diff
@@ -31,33 +36,60 @@ RL Modules (Alpha)
...
+
+We strongly recommend that neither :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_inference`
+nor :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_exploration` should be
+concerned with computing such loss-function-only data and should solely focus on computing actions alone
+(or action- logits/distribution inputs).
```
Comment (simonsays1980) [inline doc/source/rllib/rllib-rlmodule.rst:91]: Maybe adding some info about how to override these methods, namely only overriding `_forward_x` methods?

File: doc/source/rllib/rllib-rlmodule.rst [hunk 5]
```diff
@@ -408,7 +440,9 @@ There are two possible ways to extend existing RL Modules:
 Checkpointing RL Modules
 ------------------------
 
-RL Modules can be checkpointed with their two methods :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.save_to_checkpoint` and :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.from_checkpoint`.
+RL Modules can be checkpointed via their :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.save` method
+and can be constructed from a checkpoint via the :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.from_checkpoint` method.
+
```
Comment (simonsays1980) [inline doc/source/rllib/rllib-rlmodule.rst:445]: Maybe adding that `RLModule`s can be used as single instances apart of the rest of RLlib?

File: rllib/core/learner/learner_group.py [hunk 6]
```diff
@@ -625,96 +622,93 @@ def foreach_learner(
...
-                    w.load_state(path)
-                else:
-                    with tempfile.TemporaryDirectory() as temp_dir:
-                        sync_dir_between_nodes(
-                            head_node_ip, path, worker_node_ip, temp_dir
-                        )
-                        w.load_state(temp_dir)
-
-            self._worker_manager.foreach_actor(_load_state, remote_actor_ids=workers)
+    # def save_state(self, path: str) -> None:
```
Comment (simonsays1980) [inline rllib/core/learner/learner_group.py:625]: Should this be removed?

File: rllib/core/learner/torch/torch_learner.py [hunk 7]
```diff
@@ -233,7 +227,7 @@ def add_module(
         )
 
         # we need to ddpify the module that was just added to the pool
-        module = self._module[module_id]
+        module = self.module[module_id]
```
Comment (simonsays1980) [inline rllib/core/learner/torch/torch_learner.py:230]: Why do we make the module now public? To set the state explicitly?

**Review by simonsays1980 — APPROVED** [review]
LGTM. Some proposed changes for the docs and questions.

**Review by aslonnie — COMMENTED** [review]
leaving this for @simonsays1980 for now, let me know if need my action for docs approval.

**Pull request closed without merging**